### ISSUE TITLE
`-Z verbose` causes link failure

### ISSUE THREAD

**Issue by ChrisBuchholz** [rust-lang/rust#39031]
I stumbled upon a linking error when including `println!()` or `format!()` in functions that would be run by tests in a large project I'm working on. Distilled down to a short reproducible code sample, we have this:

```rust
#![feature(conservative_impl_trait)]

pub enum Error {
    Invalid,
}

pub type Result = std::result::Result<(), Error>;

/// ```
/// extern crate errr;
///
/// use errr::{runner, run};
///
/// fn main() {
///     let _ = runner(&"test test".to_string(), &run(5));
/// }
/// ```
pub fn runner<T>(subject: &T, to_run: &Fn(&T) -> Result) -> Result {
    if let Err(error) = to_run(subject) {
        Err(error)
    } else {
        Ok(())
    }
}

pub fn run(min: usize) -> impl Fn(&String) -> Result {
    move |s: &String| {
        if s.len() >= min {
            Ok(())
        } else {
            println!("test test");
            Err(Error::Invalid)
        }
    }
}

fn main() {
    let _ = runner(&"test test".to_string(), &run(5));
    println!("TESTING YEAH");
}
```

I have also put it up as a cargo project for testability here: https://github.com/ChrisBuchholz/errr-test/blob/master/src/lib.rs

Running `cargo build` as well as `rustc src/lib.rs` succeeds, whereas `cargo test` and `rustc -Z verbose src/lib.rs` errors. If I simply remove the `println!` statement from line 31, the errors disappear and everything works.

This is the output from `rustc -Z verbose src/lib.rs`:

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "lib.0.o" "-o" "lib" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd-a260b5db713b337f.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librand-664091cbac310259.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcollections-4e19d5a43d7fdd2c.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd_unicode-49cd3c7af2b2f27f.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-18bf5d50673f1daa.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunwind-6b24ec54aa474d14.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc-99e0cdfb2e11773a.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc_jemalloc-7b14c3e63843fe84.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liblibc-109501e572ed7296.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcore-f1125930d2c15bcd.rlib" "/Users/cb/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-594db3b2ae45efeb.rlib" "-l" "System" "-l" "pthread" "-l" "c" "-l" "m"
  = note: Undefined symbols for architecture x86_64:
  "_$LT$collections..string..String$u20$as$u20$core..convert..From$LT$$RF$ReEarlyBound$LP$0$C$$u20$$u27$a$RP$$u20$str$GT$$GT$::from::h4e1ae202dd816f18", referenced from:
      _$LT$str$u20$as$u20$collections..string..ToString$GT$::to_string::h36963c9d20b8cab8 in lib.0.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

error: aborting due to previous error
```

rustc version: `1.16.0-nightly (47c8d9fdc 2017-01-08)`
cargo version: `0.17.0-nightly (40a4ce6 2017-01-06)`

It's not a bug introduced in these versions though. I can confirm it has at least been a bug as far back as cargo version `nightly-2016-12-16-x86_64-apple-darwin`.

**Comment by ChrisBuchholz** [comment]
It seems that removing the `println!` statement only makes things work with `cargo test`. It still fails if I run `rustc -Z verbose src/lib.rs`.

**Comment by Nemo157** [comment]
I _think_ I've found a related bug that's triggered without using `-Z verbose`. Running `cargo build` in https://github.com/Nemo157/failing-impl-trait-thing gives

```
error: linking with `cc` failed: exit code: 1
  |
  = note: Undefined symbols for architecture x86_64:
  "foo::Foo::bar::_$u7b$$u7b$closure$u7d$$u7d$::__STATIC_FMTSTR::h3ffbe647d79e076b", referenced from:
      foo::Foo::bar::_$u7b$$u7b$closure$u7d$$u7d$::h16c823bb546d743d in temp-42fcb52dc4a86317.0.o
ld: symbol(s) not found for architecture x86_64
```

it looks like `impl trait` is somehow hiding symbol references and causing them to be stripped too early.

Interestingly this only occurs when the code is split across 2 crates, moving the subcrate into a module instead compiles fine.

Tested with:

* `rustc 1.16.0-nightly (6f1ae663e 2017-01-06)`
* `rustc 1.15.0-nightly (d9bdc636d 2016-11-24)`

**Comment by insaneinside** [comment]
@Nemo157, here we get a similar error (undefined `__STATIC_FMTSTR`) from `cargo test` -- there may be multiple bugs exposed in the test crate Chris linked.

**Comment by jyn514** [comment]
`rustc -Zverbose` works fine for me on the sample you posted - are you still running into this?

@Nemo157's crate also works fine for me as long as I change the futures dependency to 0.1.

**Comment by Nemo157** [comment]
Oh wow, ping on a very old ticket 😀

It seems very likely this was fixed in between 1.16 and the stabilization of `-> impl Trait` in 1.26, otherwise there would be a lot more bug reports about it.
